Added self handling commands

mcuyar opened the pull request:
Added self handling commands to bus

mcuyar committed:
Added self handling commands:
- Added self handler contract
- Added handler proxy to resolve self executing commands
- Updated resolver with self handling case
- Updated tests

mcuyar committed:
Refactor self handling commands
- Fixed resolver to check command instance and not handler
- Updated contract from SelfHandler to SelfHandling
- Updated handler to SelfHandlingHandler
- Updated unit tests

stevenwadejr commented:
Why does the feature need a new `SelfHandling` interface and class? Couldn't a command just implement both the `Command` and `Handler` interfaces? 

```
class FooCommand implements Command, Handler
{
    public function handle(Command $command)
    {

    }
}
```

Then the `DefaultResolver` could add:

```
case $command instanceof Command:
    return $command;
```

The Bus (soon to be Dispatcher) would then basically just be calling `handle` and passing the method it's self.

Is there more benefit to the way you did it here? Just asking.

mcuyar commented:
I originally went that route, but it felt a little dirty passing back the command to the command. It's definitely much simpler of a solution and a lot less code. It was the only way to force the handle command not to require any arguments.

I can definitely go back that route if you feel its better that way.

stevenwadejr commented:
@davidstanley01 what do you think about this approach vs the command itself implementing `Handler` as per my comment above?

davidstanley01 commented:
Sorry, I'm just digging in here... 

First off, thanks for the contribution! So glad to see that someone finds this useful outside of our org. 

I _do_ like the idea of self-handling commands. I also like the idea of a separate `SelfHandling` interface for commands.

I apologize if this is a dumb question, but why wouldn't we just check if the command is self handling in `DefaultCommandBus@execute` method and bypass the resolvers entirely?  Maybe something like this?

``` php

// DefaultCommandBus
public function execute(Command $command, $handler = null)
{
    // no need to jump through hoops finding a handler if it doesn't need one
    if ($command instanceOf SelfHandling) {
        return $command->handle();
    } 

    /**
     * The correct handler for the given command
     *
     * @var Handler $handler
     */
    $handler = $this->resolver->resolve($command, $handler);

    return $handler->handle($command);
}
```

We'd probably only need one more test to satisfy this additional condition.

stevenwadejr commented:
@davidstanley01 

![](https://i.imgflip.com/165fv5.jpg)

It's not the Bus' responsibility to figure out the handler, that's the Resolver's. If we stick with what Matt's done, it's best to keep that logic within the `DefaultResolver`, although I'd move it to the top of the `case` stack.

mcuyar commented:
David and Steven, you both make make great points. Don't know why I didn't see it before. In the `DefaultResolver` I can just switch to 

``` php
case $command instanceof SelfHandling:
    return $command->handler();
```

And everything is simple. Like Steven said, I can just move it to the top.

stevenwadejr commented:
Ok, I'm going to weigh in here again. I don't like having a separate `SelfHandlingHandler` class. I'm down for a `SelfHandling` interface that is in name only and extends the `Handler` interface. Having a separate handler class doesn't make sense here. The only reason we have a `ClosureHandler` class is to ensure we have a `handle` method, but in a `SelfHandling` class, it would be a true object and implement that method.

I also prefer

``` php
case $command instanceof SelfHandling:
    return $command;
```

That way it's simple and concise.

stevenwadejr commented:
Also, since your fake handler test function is a `noop`, will it fully cover this test scenario instead of testing that it actually returns something?

``` php
self::assertSame(
    $selfHandler->handle($command),
    $command->handle()
);
```

mcuyar committed:
Updated SelfHandling Interface to implement Handler Interface
- Removed uneeded classes
- Updated tests

mcuyar commented:
Based on feedback, I've updated the pull request where the `SelfHandling` interface now extends the Handler interface.

The reason the `SelfHandlingHandler` existed was to avoid passing the `Command` class back to itself. It is now removed and the handle method now requires the `Command` instance as a parameter.
